skaha/src/main/java/org/opencadc/skaha/image/GetAction.java

```diff
@@ -102,30 +98,35 @@ public GetAction() {
     @Override
     public void doAction() throws Exception {
         super.initRequest();
-        
+
         String type = syncInput.getParameter("type");
         if (type != null && !SESSION_TYPES.contains(type)) {
             throw new IllegalArgumentException("unknown type: " + type);
         }
-        
+
         String idToken = super.getIdToken();
         List<Image> images = getImages(idToken, type);
         Gson gson = new GsonBuilder().disableHtmlEscaping().setPrettyPrinting().create();
         String json = gson.toJson(images);
-        
+
         syncOutput.setHeader("Content-Type", "application/json");
         syncOutput.getOutputStream().write(json.getBytes());
     }
-        
+
     protected List<Image> getImages(String idToken, String typeFilter) throws Exception {
+        String lastUpdated = redis.get(IMAGE_LAST_UPDATED_TIMESTAMP);
+        if (isImageListUpdated(lastUpdated)) {
+            return redis.get(PUBLIC_IMAGES, ImageList.class).get();
+        }
+
```

Comment:
I think there is a threading and timing issue with the cache refresh.  If you take our current production deployment as an example, the call(s) to list images from harbor takes quite a while right now, let's say 10 seconds.  During this 10 seconds, more requests to list images will come in and will also attempt to refresh the cache because the timestamp has not yet been updated.  So, one issue is that there should only be 1 thread trying to refresh the cache at any given time.

Also, by using the user's thread to do the cache refresh, it will result in an inconsistent user experience.  Some calls may use the cache and be really fast and others won't.

So it's sounding like the cache refresh should be done in the backgroud.  Maybe a k8s cron is the way to do manage the cache?  Then skaha would use the cache on each call.  However you then have a startup dependency--the cron must run once before skaha can use the cache.

Open to ideas on this, but changes are definitely needed.